report_backend/time_spent_report.py

```diff
@@ -0,0 +1,138 @@
+"""
+Time spent report backend.
+"""
+import csv
+from datetime import datetime
+import os
+
+import boto3
+
+
+from .base import AbstractBaseReportBackend
+
+
+class TimeSpentReportBackend(AbstractBaseReportBackend):
+    """
+    Backend for time spent report.
+    """
+
+    def __init__(self, **kwargs):
+        super(TimeSpentReportBackend, self).__init__(**kwargs)
+
+
+    def json_report_to_csv(self, json_report_data):
+        """
+        Process json data to convert into csv format.
+
+        Args:
+            json_report_data: Json representacion containing time spent report data.
+        """
+        report_data = json_report_data.get('result', {})
+
+        if not report_data:
+            print('No report data...')
+            exit()
+
+        if report_data.get('time_spent_data'):
+            time_spent_data = report_data.get('time_spent_data')
+            for course in time_spent_data.keys():
+                file_name = '{}'.format(course)
+
+                if not time_spent_data.get(course):
+                    continue
+
+                analytics_data = time_spent_data.get(course).get('analytics_data', [])
+                course_structure_data = time_spent_data.get(course).get('course_structure', [])
+
+                subsection_data = count_analytics_subsections(analytics_data, course_structure_data)
+                report_csv_headers = subsection_data[0].keys()
+
+                self.create_csv_file(file_name, subsection_data, report_csv_headers)
+
+
+    def create_csv_file(self, file_name, body_dict, headers):
+        """
+        Creates the csv file with the passed arguments, and then save it locally.
+
+        Args:
+            file_name: File string name.
+            body_dict: Dict with the data to write the csv file.
+            headers: List with the csv column names.
+        """
+        path_file = '{parent_folder}/result/{file_name}.csv'.format(
+            parent_folder=os.path.join(os.path.dirname(__file__), os.pardir),
+            file_name=file_name
+        )
+
+        with open(path_file, mode='w', encoding='utf-8') as csv_file:
+            column_headers = headers
+            writer = csv.DictWriter(csv_file, fieldnames=column_headers)
+
+            writer.writeheader()
+            for row in body_dict:
+                writer.writerow(row)
+
+        self.upload_file_to_storage(file_name, path_file)
+
+
+    def upload_file_to_storage(self, course, path_file):
```

Comment:
This method is the same  for completion and last page, I think this should be move to a common file in order to avoid to re-define it

Reply:
You right, but it could be implemented at the end of reports, since we need to redefine the way that the reports are saving in S3. @andrey-canon